File: OJ/Leetcode/Algorithm/325. Maximum Size Subarray Sum Equals k.java
public class Solution { // 29ms  
    public int maxSubArrayLen(int[] nums, int k) {  
        int sum = 0;  
        HashMap<Integer, Integer> map = new HashMap<>(); //key:prefix和; value:到第几个数  
        int max = 0;  
        map.put(0, 0);  
        for (int i = 1; i < nums.length + 1; i++) {  
            sum += nums[i - 1];  
            int lookingfor = sum - k;  
            if (map.containsKey(lookingfor)) {  
                max = Math.max(max, i - map.get(lookingfor));  
            }  
            if (!map.containsKey(sum))  
                map.put(sum, i);  
        }  
        return max;  
    }  
}  

File: OJ/Leetcode/Algorithm/325. Maximum Size Subarray Sum Equals k.cpp
class Solution {
public:

    int maxSubArrayLen(vector<int>& nums, int k) {
		unordered_map<int, int> mp; // sum->pos
		int n = nums.size(), sum = 0, ans = 0;
		mp[0] = -1; // for the situation of 0~i
		for (int i = 0; i < n; i++){
			sum += nums[i];
			if (mp.find(sum-k) != mp.end())
				ans = max(i - mp[sum-k], ans);
			if (mp.find(sum) == mp.end())
				mp[sum] = i;
		}
		return ans;
    }
};
